File: markdown/src/main/java/ru/noties/markwon/extension/emoji/internal/EmojiDelimiterProcessor.java
package ru.noties.markwon.extension.emoji.internal;

import org.commonmark.node.Node;
import org.commonmark.node.Text;
import org.commonmark.parser.delimiter.DelimiterProcessor;
import org.commonmark.parser.delimiter.DelimiterRun;

import ru.noties.markwon.extension.emoji.Emoji;

public class EmojiDelimiterProcessor implements DelimiterProcessor {

    @Override public char getOpeningCharacter() {
        return ':';
    }

    @Override public char getClosingCharacter() {
        return ':';
    }

    @Override public int getMinLength() {
        return 1;
    }

    @Override public int getDelimiterUse(DelimiterRun opener, DelimiterRun closer) {
        if (opener.length() >= 1 && closer.length() >= 1) {
            return 1;
        } else {
            return 0;
        }
    }

    @Override public void process(Text opener, Text closer, int delimiterCount) {
        Emoji emoji = null;
        Node text = opener.getNext();
        if (text instanceof Text) {
            emoji = new Emoji();
            emoji.setEmoji(((Text) text).getLiteral());
            text.unlink();
        }
        if (emoji != null) opener.insertAfter(emoji);
    }
}

File: markdown/src/main/java/com/fastaccess/markdown/extension/markwon/emoji/internal/EmojiDelimiterProcessor.kt
package com.fastaccess.markdown.extension.markwon.emoji.internal


import com.fastaccess.markdown.extension.markwon.emoji.Emoji
import org.commonmark.node.Text
import org.commonmark.parser.delimiter.DelimiterProcessor
import org.commonmark.parser.delimiter.DelimiterRun

class EmojiDelimiterProcessor : DelimiterProcessor {

    override fun getOpeningCharacter(): Char = ':'

    override fun getClosingCharacter(): Char = ':'

    override fun getMinLength(): Int = 1

    override fun getDelimiterUse(
        opener: DelimiterRun,
        closer: DelimiterRun
    ): Int = if (opener.length() >= 1 && closer.length() >= 1) {
        1
    } else {
        0
    }

    override fun process(
        opener: Text,
        closer: Text,
        delimiterCount: Int
    ) {
        var emoji: Emoji? = null
        val text = opener.next
        if (text is Text) {
            emoji = Emoji()
            emoji.emoji = text.literal
            text.unlink()
        }
        if (emoji != null) opener.insertAfter(emoji)
    }
}
